Instrumentation into Block IR

ChingLongTin opened the pull request:
Call implementation will be after the algorithms for instrumenting functions are completed.

1. allocating `optionNme`
This changed some diff tests (`OpenWildcard.mls`, `Imports.mls`, `ImportMLs.mls`) that imported Option themselves.
If I [allocate `optionNme` after checking if `stageCode` is enabled](https://github.com/hkust-taco/mlscript/commit/aae0086c576f3af053909ef64b357de11d27bb0c), it fails to link to the file.
Is there some way to avoid changing the diff tests, or is this fine?

2. implementing I-Inst
The current shape for Class `C(n:s)` includes the field names (`n`), but the field names are not available in `Instantiation`.
The field names are used in `sel`. If the class is staged using I-Cls, then we can use the staged `ClsLikeDefn` to retrieve the symbol names, but when do we do when it isn't staged?
```
class A(a)
staged module B
  class C(a)
  fun f() = 
    new C(1) // ok, C is staged
    new A(1) // not ok, we don't know field names of A after instrumentation
```

ChingLongTin committed:
Update Syntax.mls

ChingLongTin committed:
move function

ChingLongTin committed:
implement more rules

ChingLongTin committed:
add  stubs for function instrumentation

ChingLongTin committed:
add instrumentation context

ChingLongTin committed:
refactor transformArgs

ChingLongTin committed:
add Context to some functions

ChingLongTin committed:
dedup Pattern with Block.Case

ChingLongTin committed:
rename mls{Block, Shape, Symbol} helpers

ChingLongTin committed:
amend! link to functions in Shape.mls

link to functions in Shape.mls

ChingLongTin committed:
flesh out {Block, Shape}.mls

ChingLongTin committed:
formatting

ChingLongTin committed:
implement ruleMatch

ChingLongTin committed:
remove Label from Block.mls

ChingLongTin committed:
rename helpers

ChingLongTin committed:
update Block.mls

ChingLongTin committed:
adapt NeilKleistGao's example code

ChingLongTin committed:
rename helpers

ChingLongTin committed:
refactor ruleMatch

ChingLongTin committed:
fix ruleTup

ChingLongTin committed:
reduce supported cases to make mvp

ChingLongTin committed:
move detecting staged annotation to ClsLikeDefn

ChingLongTin committed:
move importing to diff testing compiler flag

ChingLongTin committed:
revert formatting changes

ChingLongTin committed:
Update Printer.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

ChingLongTin committed:
Update Lowering.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

ChingLongTin committed:
Merge branch 'hkmc2' into staged-syntax

ChingLongTin committed:
remove redundant :pt

ChingLongTin committed:
fix syntax error

ChingLongTin committed:
fix type hint for arms in Match

yeungsinchun committed:
add preliminary printCode function

ChingLongTin committed:
move instrumentation to JSBackendDiffMaker

ChingLongTin committed:
rename compiler flag to :staging

ChingLongTin committed:
Revert "remove redundant :pt"

This reverts commit 317c6f6c3a2ae963d064f1ad3b011da396bc1dfe.

ChingLongTin committed:
combine import statements to list

ChingLongTin committed:
Merge branch 'staged-syntax' into instrument

ChingLongTin committed:
reword function

ChingLongTin committed:
Revert "move instrumentation to JSBackendDiffMaker"

This reverts commit e4d7b227ee3cab3d5a7461ec746cf23468b5ae8d.

ChingLongTin committed:
Merge branch 'staged-syntax' into instrument-simple

ChingLongTin committed:
prune implementation for mvp

ChingLongTin committed:
implement rudimentary instrumentation logic

ChingLongTin committed:
implement printing for DynSelect

ChingLongTin committed:
only stage functions in staged modules

ChingLongTin committed:
update Syntax.mls

yeungsinchun committed:
remove bugs in pretty printing

yeungsinchun committed:
Merge branch 'pretty-print' into instrument-simple

ChingLongTin committed:
formatting

ChingLongTin committed:
add back Symbol class
this was done to be in line with Block, though there's no reason Symbol can't just be a Str

ChingLongTin committed:
reuse original function name

ChingLongTin committed:
update test cases

ChingLongTin committed:
fix ruleVal

ChingLongTin committed:
add some debug symbol names

ChingLongTin committed:
rename PathLike to ArgWrappable

ChingLongTin committed:
fix applyBlock impl
this makes it so modules that are not the first node of the Block will still be staged

ChingLongTin committed:
refactor transformFunDefn

clean up function staging logic

ChingLongTin committed:
remove spread from Arg
this makes having to import Option and makes transformArg easier to implement

ChingLongTin committed:
clean up outdated comments and helpers

ChingLongTin committed:
implement ruleTup

ChingLongTin committed:
formatting

ChingLongTin committed:
implement ruleSel and ruleInst

ChingLongTin committed:
remove outdated comment

ChingLongTin committed:
clean up symName printing

ChingLongTin committed:
formatting

yeungsinchun committed:
Add shape in diff testing

ChingLongTin committed:
update note
as BlockTransformer doesn't pass around a context, the implementation should be more than just a extension of BlockTransformer

ChingLongTin committed:
update PrintCode printout

ChingLongTin committed:
update Shape constructors

ChingLongTin committed:
add context

ChingLongTin committed:
remove outdated comments

yeungsinchun committed:
correc comment

yeungsinchun committed:
fix pretty print to print mlscript

ChingLongTin committed:
refactor
add back symName parameter to {block, shape}Ctor

ChingLongTin committed:
follow JSBuilder pretty printing for DynSelect in Printer

ChingLongTin committed:
remove Field from Case
this makes Block.Case line up with formalization of Pattern

ChingLongTin committed:
change transformBlock signature

ChingLongTin committed:
implement more rules

ChingLongTin committed:
remove import Option from Shape.mls

ChingLongTin committed:
replace Multiple with ShapeSet
we keep ShapeSet and Shape separate for now

ChingLongTin committed:
misc fixes to Shape.mls

ChingLongTin committed:
patch showLiteral for null and undefined

ChingLongTin committed:
replace ruleVal

ChingLongTin committed:
clean up

ChingLongTin committed:
add symbol to Class Shape for matching

ChingLongTin committed:
add spread field to Arg
this introduces a new Option symbol to Elaborator
if paramsOpt and companion are unused for ClsLikeDefn in Block.mls, it might be possible to avoid using Option for defining Block.mls

ChingLongTin committed:
update PrintCode.mls

ChingLongTin committed:
add ruleCls

ChingLongTin committed:
formatting

ChingLongTin committed:
progress on rest of the rules
removed Wildcard from Shape, as we already store the branch in dflt

ChingLongTin committed:
replace Block.Case with Pattern
they aren't the same thing

ChingLongTin committed:
implement ShapeSet and ShapeMap

ChingLongTin committed:
fix equality checks for Symbol

ChingLongTin committed:
implement Shape{Map, Set}
use pretty printing for the key for now

ChingLongTin committed:
fill out ShapeSet functionailty

ChingLongTin committed:
link ShapeSet symbols for instrumentation

ChingLongTin committed:
fix Shape, ShapeSet implementations

ChingLongTin committed:
remove TrivialResult
this helps avoid applying TrivialResult multiple times when retrieving a path from the Context

ChingLongTin committed:
replace Shape with ShapeSet during instrumentation

ChingLongTin committed:
add message to assertions

ChingLongTin committed:
refactor match rules

ChingLongTin committed:
add debug print Shape

ChingLongTin committed:
print more information for ClsLikeDefn

ChingLongTin committed:
combine Pattern with Case
we transform Some(dflt) branch in a Match statement without calling the filter function at JS runtime

ChingLongTin committed:
fill out more match instrumentation

ChingLongTin committed:
update comments

ChingLongTin committed:
update PrintCode.mls

ChingLongTin committed:
move optionNme inside stageCode scope
this avoids renaming Option in test files that import Option separately, see ImportMLS.mls, Imports.mls, OpenWildcard.mls

ChingLongTin committed:
Merge remote-tracking branch 'mlscript/hkmc2' into instrument

ChingLongTin committed:
Revert "move optionNme inside stageCode scope"

This reverts commit c90118c54ac819133d4ae44016714730c2d1ae0f.
this leads to Option not being available, updating the diff tests affected by the renaming

ChingLongTin committed:
use more meaningful names in Functions.mls

ChingLongTin committed:
refactor collectApply

ChingLongTin committed:
formatting

ChingLongTin committed:
update comments and debug info

ChingLongTin committed:
remove redundant import

ChingLongTin committed:
match pretty printing syntax for DynSelect with formalization

ChingLongTin committed:
fix mkClass

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
formatting

ChingLongTin committed:
Merge branch 'instrument' of https://github.com/ChingLongTin/mlscript into instrument

LPTK commented:
> * allocating `optionNme`
>   This changed some diff tests (`OpenWildcard.mls`, `Imports.mls`, `ImportMLs.mls`) that imported Option themselves.
>   If I [allocate `optionNme` after checking if `stageCode` is enabled](https://github.com/hkust-taco/mlscript/commit/aae0086c576f3af053909ef64b357de11d27bb0c), it fails to link to the file.
>   Is there some way to avoid changing the diff tests, or is this fine?
> *

Just follow the pattern of `Runtime`: you can name the symbols
```scala
    val shapeSetSymbol = TempSymbol(N, "shapeSet")
    val optionSymbol = TempSymbol(N, "option")
```
to avoid unnecessarily reserving the name `Option`.

ChingLongTin committed:
fix syntax errors

the staged modifier is lost somewhere during lowering

ChingLongTin committed:
patch in instrumentation

temporarily stage all modules until i can find where the staged modifier went

ChingLongTin committed:
unreserve Option in Elaborator

LPTK commented:
> 2\. The current shape for Class `C(n:s)` includes the field names (`n`), but the field names are not available in `Instantiation`.
>     The field names are used in `sel`. If the class is staged using I-Cls, then we can use the staged `ClsLikeDefn` to retrieve the symbol names, but when do we do when it isn't staged?

So, in principle it should be possible to obtain information about a class from its symbol. Could we make the symbols referring to classes point to the class' `ClsLikeDefn`, even if that class is not staged (in which case we can probably store less information in the `ClsLikeDefn`)?

ChingLongTin committed:
remove argument names from class shape

ChingLongTin committed:
move functions

ChingLongTin committed:
add ClassSymbol

ChingLongTin committed:
formatting

ChingLongTin committed:
clean up

ChingLongTin committed:
Revert "move functions"

This reverts commit 1f5c8d7125c3735f3271e3551daac4bf75c0db53.

ChingLongTin committed:
rename functions

ChingLongTin committed:
add debug information to rules

ChingLongTin committed:
move paramsOpt into ClassSymbol

yeungsinchun committed:
update diff output

ChingLongTin committed:
implement pruning of bad arms from Match statement

yeungsinchun committed:
pass staged annotation in lowering

ChingLongTin committed:
add debug name to transformSymbol

ChingLongTin committed:
remove code duplication from ruleBranch

ChingLongTin committed:
fix typo

ChingLongTin committed:
Revert "patch in instrumentation"

This reverts commit 3c4faec660df12baae4a93f8abdb958c1034007e.

LPTK commented on hkmc2/shared/src/test/mlscript-compile/Block.mls:
```diff
@@ -29,7 +29,7 @@ class Case with
   constructor
     Lit(val lit: Literal)
     Cls(val cls: Symbol, val path: Path)
-    Tup(val len: Int) // TODO: remove inf?
```
FYI `inf = true` is suppossed to indicate that the `len` is a lower bound and that the tuple could be bigger. It's what you get from a pattern like `[0, 1, ...xs, 8, 9]`.

ChingLongTin replied:
Do we add it back? In the formailzation the match pattern `[]^n` seems to indicate we only handle cases where `n` is exact.

LPTK replied:
It's better to stick closest to the real IR. The formalization is slightly "idealized" to simplify the presentation.

LPTK commented:
Do you really need `showDefn`? I think the default pretty-printing is probably better, as it lays out large values on multiple lines.

ChingLongTin commented:
> Do you really need `showDefn`? I think the default pretty-printing is probably better, as it lays out large values on multiple lines.

The `show` functions are currently a placeholder for later, when we need to convert the mlscript Block constructs back into mlscript in [this branch](https://github.com/ChingLongTin/mlscript/tree/fix-pretty-print). For now, it just has the effect of ensuring the expression is well-typed.

ChingLongTin commented:
> Could we make the symbols referring to classes point to the class' `ClsLikeDefn`?

I have added a `ClassSymbol` class, so that both staged and unstaged functions can access the parameter names after instrumentation. So for example in `Functions.mls`,
```
:js
:staging
:sjs
class Outside(a)
staged module ClassInstrumentation with
  class NoArg
  fun inst1() = new Outside(1)
  fun inst2() = new NoArg
```
In `inst1` and `inst2`, there will be `ClassSymbol("Outside", Some([Symbol("a"), Symbol("b")]))` and `ClassSymbol("NoArg", None))` to access the parameter names.

yeungsinchun committed:
Merge branch 'fix-pretty-print' into instrument

yeungsinchun committed:
update pretty printing to print actual mlscript

yeungsinchun committed:
only pass the "staged" annotation for non-companion module

yeungsinchun committed:
Merge remote-tracking branch 'origin/instrument' into instrument

yeungsinchun committed:
reduce test cases

yeungsinchun committed:
fix class definition printing

yeungsinchun committed:
add match error on pretty printing

yeungsinchun committed:
fix test cases

yeungsinchun committed:
some unimportant change

yeungsinchun committed:
some more unimportant change

yeungsinchun committed:
Merge branch 'instrument-with-pretty-print' into instrument

yeungsinchun committed:
printed correctly now

ChingLongTin committed:
add inf back to Tup

still need a test case for it

yeungsinchun committed:
Revert "printed correctly now"

This reverts commit f582fa5119763e1d66484b6959d9581de63b58a7.

yeungsinchun committed:
Revert "Merge branch 'instrument-with-pretty-print' into instrument"

This reverts commit 2548c78417bb38a3be76ecc43579a0365cb6fb60, reversing
changes made to 7151735a72a4b7dd86217743fe2c96191e912d9b.

yeungsinchun committed:
Revert "some more unimportant change"

This reverts commit 7151735a72a4b7dd86217743fe2c96191e912d9b.

yeungsinchun committed:
Revert "Merge remote-tracking branch 'origin/instrument' into instrument"

This reverts commit 2b1d1576279f2d8aca058036f8671f6d92deb161, reversing
changes made to cc24cae44e234f6f43c8c3638ad754a1a3ca3742.

yeungsinchun committed:
Revert "update pretty printing to print actual mlscript"

This reverts commit 5d63091e0d1e91de74d0601fc8bf2ad5155f2c59.

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
fix showArg printing

ChingLongTin committed:
add more Tup test cases

ChingLongTin committed:
formatting

ChingLongTin committed:
Reapply "Merge remote-tracking branch 'origin/instrument' into instrument"

This reverts commit a3925e5da38dfbb0b92d9a0ddf59aee45ccabedf.

ChingLongTin committed:
fix static impl

ChingLongTin committed:
fix params type in FunDefn

LPTK commented on hkmc2/shared/src/test/mlscript-compile/Shape.mls:
```diff
@@ -28,31 +28,32 @@ class Shape with
-    Lit(l) and s2 is Class(sym, params)
+fun mrg2(s1: Shape, s2: Shape) =
+  // we want different objects with same parameters to be equal
+  if show(s1) == show(s2) then s1
```
This is a huge temporary hack, and deserves at least a big
```suggestion
  if show(s1) == show(s2) then s1 // FIXME: hack!
```

LPTK replied:
An interesting side project you could undertake is to add generation of equality functions for `data class`es.

ChingLongTin committed:
add Context to ruleEnd

ChingLongTin committed:
uppdate Match rules

this implements the rest function that narrows the shape as more match statements are processed

ChingLongTin committed:
mark pretty printing equality as fixme

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
progress: deprecate ShapeSet from StagedPath

ChingLongTin committed:
progress: strip shape logic from rules

ChingLongTin committed:
remove inferring shape from Instrumentation

as discussed in the meeting, we will perform the within mlscript to reduce complexity

ChingLongTin committed:
progress: add StagedPath to ArgWrappable

ChingLongTin committed:
progress: simplify impl

ChingLongTin committed:
progress: formatting

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin commented:
From the meeting, the logic for shape will now be implemented in mlsript, so I am reducing the scope of Instrumentation.scala to just handle conversion from Block into Block IR. I will mark this pr as ready when that is done.

ChingLongTin committed:
Merge remote-tracking branch 'mlscript/hkmc2' into instrument

ChingLongTin committed:
fix printCode syntax

ChingLongTin committed:
remove equality hack in Shape.mls

ChingLongTin committed:
switch from ValDefn to Scoped

ChingLongTin committed:
fix: make new dSym when cloning FunDefn

ChingLongTin committed:
remove context from ruleEnd

ChingLongTin committed:
add staging for functions with arguments

ChingLongTin committed:
progress: simplify StagedPath

ChingLongTin committed:
progress: Make blockCtor produce StagedPath

ChingLongTin committed:
use ClassSymbol for ruleApp

ChingLongTin committed:
formatting

ChingLongTin committed:
update tests

ChingLongTin committed:
remove StagedPath

after shape information is removed, the extra typechecking is no longer helpful

ChingLongTin committed:
formatting

ChingLongTin committed:
inline functions

ChingLongTin committed:
formatting

use no ident after match, like in Lowering.scala

ChingLongTin committed:
remove Shape{Map, Set} definitions

the definitions should utilize CachedHash before getting merged

ChingLongTin committed:
formatting

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
formatting

ChingLongTin committed:
imrpove symbol information for select

ChingLongTin committed:
simplify applyRuleBranch

since the shape is not inferred, each branch can use the same x to compute its branch

ChingLongTin committed:
import option file during MLsDiffMaker init

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/MLsDiffMaker.scala:
```diff
@@ -167,7 +167,9 @@ abstract class MLsDiffMaker extends DiffMaker:
       processTrees(
-        PrefixApp(Keywrd(`import`), StrLit(blockFile.toString)) :: Nil)
+        PrefixApp(Keywrd(`import`), StrLit(blockFile.toString))
+        :: PrefixApp(Keywrd(`import`), StrLit(optionFile.toString))
```
Such imports are for things meant to be accessed by _users_, as they will lead to the introduction of a new name in scope (here, `Option`). But this won't work that well for users as you're importing the `mjs` instead of the `mls`, so we won't get MLscript symbols.

I think the import you want here is closer to the way `Runtime` is imported, which is never made available. Check out how that's done.

For the record, it looks like the `termFile` import above added by @NeilKleistGao is equally wrong.

ChingLongTin replied:
I think you mean like in `JSBackendDiffMaker.scala:58` using `importRuntimeModule`, which is already replicated for both `Block` and `Option` in line 63-64.

I removed the import statement for `Option` in `MLsDiffMaker`, and kept the one for `Block`, since the former isn't available, while the latter is.

LPTK replied:
Why is Block being handled differently than Option and Term? And again, if you want this to behave well, you should import the `mls`, not `mjs`.

ChingLongTin replied:
ah, so what you meant was to just never make Block and Option available to the user. If we need the MLscript symbols, we would need to directly import the mls file in the mls file instead, instead of in the diff maker's `init`.

LPTK replied:
Yes, more or less. Things that are imported by default are defined in `Predef`. There is no need to import more things by default. And if we did want to import these additional things, that should be done correctly by importing the `.mls` and not the `.mjs`.

ChingLongTin committed:
remove import statement for option in MLsDiffMaker

ChingLongTin committed:
include staging tests in diff tests

renaming mlsript/staging to mlscript/block-staging avoids DiffTestRunner excluding the test files using the filter intended to exlude ucs/staging

ChingLongTin committed:
fixup! remove import statement for option in MLsDiffMaker

ChingLongTin committed:
remove redundant codegen

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
add Block symbol following pr #375

ChingLongTin committed:
formatting

ChingLongTin committed:
add Scoped, Loop, Break nodes to Block IR

this removes the check in L218 because evaluation of Scope is now deferred:wq

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala:
```diff
@@ -310,6 +310,7 @@ object Elaborator:
     def init(using State): Ctx = Ctx.empty.copy(env = Map(
       "globalThis" -> globalThisSymbol,
       "Term" -> termSymbol,
+      "Block" -> blockSymbol,
```
Again, I don't think this should be here and neither should `"Term"`. This is the context meant for the user. For internal uses, one should use an internal symbol (like `runtimeSymbol`), not a plain name. Cc @NeilKleistGao.

NeilKleistGao replied:
I think the `block` and the `shape` module would not be used explicitly by users, but users do need to write code like `Term.codegen`. if we remove it from the context, then users should import the `term` module explicitly.

LPTK replied:
If some functions like this should be accessible by users, they should be in the `Predef`. The `Predef` could either re-export `Term` (but it's a bit weird, since most things in it are not meant to be used directly by users) or define some submodule, such as `module meta with { fun codegen... }`.

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
replace parameters of functions with its symbol instead of undefined

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
retain scoped symbols after instrumentation

ChingLongTin committed:
update test

ChingLongTin committed:
replace calls to staged functions

ChingLongTin committed:
Wrap function body

ChingLongTin committed:
move printing into constructor

ChingLongTin committed:
remove _gen suffix from instrumented block

we check if a function is staged through defCtx, so we can substitute the symbol there

ChingLongTin committed:
remove parameters from instrumentation function

_instr function only returns the block IR, whereas _gen functions will include the shape propagation

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

NeilKleistGao commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Instrumentation.scala:
```diff
@@ -4,12 +4,326 @@ package codegen
+      val debugBlock = debugPrintCode.foldRight(Return(unit, true))(concat)
+      def debugCont(rest: Block) =
+        Begin(debugBlock, rest)
+      val newCtor = impl.transformBlock(companion.ctor)(using new HashMap())(_ => debugCont(End()))
```
So far, if we introduce things from #387, it will fail, since we do not consider the mutable features that #387 is using. We probably need to only handle classes inside modules and leave other things untouched.

ChingLongTin replied:
I'm not sure how to bring the classes defined in the module to the next stage, The class `C` depends on the variable `x` defined within the module, so the `let` statement needs to be staged as well.
Will the classes never depend on variables that use mutable features?
```
module A with
  let x = 1
  fun get() = x
  class C() with
    fun get() = x
  fun add() =
    x = x + 1
```

LPTK replied:
Mutation should probably not be allowed in staged modules, I think.

BTW, it should be `set x = x + 1` or equivalently `set x += 1`. OTOH `x = x + 1` is an only an initialization statement (and wll be rejected here once we have type checking).

ChingLongTin replied:
From our meeting on Tuesday, the mutable feature was only used to ensure no name clash happens for the generated class, but since the generated symbol name can't be spelled by the user, we don't need the mutable feature in the constructor. Therefore, we are able to stage the constructor normally.

ChingLongTin committed:
minor nit

NeilKleistGao commented on hkmc2/shared/src/test/mlscript-compile/Block.mls:
```diff
@@ -0,0 +1,189 @@
+  
+fun concat(b1: Block, b2: Block) = if b1 is
+  Match(scrut, arms, dflt, rest) then Match(scrut, arms, dflt, concat(rest, b2))
+  Return(res, implct) then b2 // discard return?
```
`b1` should not be discarded. When we have an early-return, we should discard the rest of the body, i.e., `b2`.

NeilKleistGao commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Instrumentation.scala:
```diff
@@ -4,12 +4,326 @@ package codegen
+        blockCtor("ClassSymbol", Ls(toValue(sym.nme), paramsOpt), symName)(k)
+    case t: TermSymbol if t.defn.exists(_.sym.asCls.isDefined) =>
+      transformSymbol(t.defn.get.sym.asCls.get, symName)(k)
+    case _ => blockCtor("Symbol", Ls(toValue(sym.nme)), symName)(k)
```
Notice that the default name of `TempSymbol` is always `tmp`, and those symbols will be renamed with a number suffix during codegen. We probably also need to use `Scope` to avoid name clashes here.

ChingLongTin committed:
rename TempSymbol when staging

ChingLongTin committed:
fix concat logic

ChingLongTin committed:
warn shape propogation doesn't cover multiple parameter lists

ChingLongTin committed:
move ctor staging to a function

ChingLongTin committed:
WIP: add class staging test

ChingLongTin committed:
resolve fixmes

ChingLongTin committed:
stage class functions

ChingLongTin committed:
Merge remote-tracking branch 'mlscript/hkmc2' into instrument

ChingLongTin committed:
Merge remote-tracking branch 'mlscript/hkmc2' into instrument

ChingLongTin committed:
update tests

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

LPTK committed:
Merge branch 'hkmc2' into instrument

LPTK commented:
Somehow there is a compilation error...

ChingLongTin committed:
Merge remote-tracking branch 'mlscript/hkmc2' into instrument

ChingLongTin committed:
Merge remote-tracking branch 'origin/instrument' into instrument

ChingLongTin committed:
always enable staging

NeilKleistGao commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Instrumentation.scala:
```diff
@@ -4,12 +4,346 @@ package codegen
+  type ArgWrappable = Path | Symbol
+  type Context = HashMap[Path, Path]
+  // TODO: there could be a fresh scope per function body, instead of a single one for the entire program
+  var scope = Scope.empty
```
```suggestion
  val scope = Scope.empty
```

NeilKleistGao commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Instrumentation.scala:
```diff
@@ -4,12 +4,346 @@ package codegen
+        transformPath(qual): x =>
+          transformPath(fld): y =>
+            blockCtor("DynSelect", Ls(x, y, toValue(arrayIdx)), "dynsel")(k)
+      case _ => ??? // not supported
```
Better to throw a compilation error to say it is not supported.

NeilKleistGao commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Instrumentation.scala:
```diff
@@ -4,12 +4,346 @@ package codegen
+        blockCtor("Break", Ls(labelSymbol))(k(_, ctx))
+    case _ => ??? // not supported
+
+  def transformFunDefn2(f: FunDefn)(using Context)(k: Path => Block): Block =
```
What's the difference between `transformFunDefn` and `transformFunDefn2`?

ChingLongTin replied:
`transformFunDefn2` is a continuation that gets reused to stage class functions. If the classes are lifted out of the module, then it can just be inlined into `transformFunDefn`.

LPTK replied:
It's fine to leave it out, but consider using a more descriptive name.

NeilKleistGao commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Instrumentation.scala:
```diff
@@ -4,12 +4,346 @@ package codegen
+  def concat(b1: Block, b2: Block): Block =
+    b1.mapTail {
+      case _: End => b2
+      case _ => ???
```
```suggestion
      case _ => lastWords("reason why this branch should be unreachable.")
```

NeilKleistGao commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Lowering.scala:
```diff
@@ -269,13 +269,28 @@ class Lowering()(using Config, TL, Raise, State, Ctx):
+                //   stagedAnnots
+                // )
+                // clsSymb.defn = S(newDefn)
+                // newDefn
```
Remove the comments if they are no longer useful.

NeilKleistGao commented on hkmc2/shared/src/test/mlscript-compile/Shape.mls:
```diff
@@ -0,0 +1,120 @@
+    [Arr(s1, false), Arr(s2, false)]
+      and s1.length == s2.length
+      then Arr(zipMrg(s1, s2), false)
+    else Dyn()
```
I think you can remove it (and other functions that will no longer be used).

NeilKleistGao commented on hkmc2/shared/src/test/mlscript-compile/Shape.mls:
```diff
@@ -0,0 +1,120 @@
+  constructor
+    Dyn()
+    Lit(val l: Literal)
+    Arr(val shapes: Array[Shape], val inf: Bool)
```
It seems that we forgot to discuss whether we really need spread parameters and infinite arrays in this project. How do you think? @LPTK

LPTK replied:
There is no such thing as an infinite array, so not sure what you mean.

Spread parameters are important, but not on the critical path. Let's try to do without spreads for now, and add them later or if they become critical for progress.

NeilKleistGao replied:
Then we shall remove this `in` parameter, I guess, and mark spread parameters not supported for now.

ChingLongTin replied:
Spread operators are also what causes `Block.Tup` to have `inf = true`, as discussed here: https://github.com/hkust-taco/mlscript/pull/355/#discussion_r2563723032

Are we also disallowing pattern matching with spread parameters, so both (1) and (2) are not allowed?
```
if [1, ..[1, 2]] is // (1: spread in Arg)
  [1, ...x] then x // (2: spread in Tup)
  else 0
```

LPTK replied:
Yes, it's probably fine to reject them _for now_. Again, ideally, we'll try to revisit this later, though.

LPTK replied:
I've added this as a draft issue in the Dynamic Staging project https://github.com/orgs/hkust-taco/projects/17/views/1?pane=issue&itemId=162809949

NeilKleistGao commented on hkmc2/shared/src/test/mlscript/block-staging/Functions.mls:
```diff
@@ -0,0 +1,114 @@
+    x = 1
+    x
+  fun g(x)(y, z)() = z
+//│ ╔══[WARNING] Multiple parameter lists are not supported in shape propagation yet.
```
Functions with multiple parameter lists can be desugared by the first-class function transformer (Enabled by `:ftc`. Probably after #404 is merged).

NeilKleistGao commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Instrumentation.scala:
```diff
@@ -4,12 +4,346 @@ package codegen
+    sym match
+    case clsSym: ClassSymbol =>
+      val name = scope.allocateOrGetName(sym)
+      transformParamsOpt(clsSym.defn.get.paramsOpt): paramsOpt =>
```
Some class symbols (e.g., generated by another pass) may have no `defn`, which will lead to a crash.

ChingLongTin replied:
Since the parameters are necessary to reconstruct the class instance during specialization, this case throws a compiler error for now.

NeilKleistGao replied:
I see. Then it is enough to raise a compilation error here. I will set a dummy definition with parameters in #404 so that our code can cooperate.

Is there any other information that you need to retrieve from the `defn` field?

ChingLongTin replied:
From our discussions last time, we need `paramsOpt` and `auxParams`.

NeilKleistGao replied:
OK. It seems that dummy definitions do not help either. I think a proper way is to have another pass before starting instrumentation to collect these locally generated classes (represented in block IR). If there is no definition for the class symbol, you try to retrieve it from the collected data. These generated classes are not visible outside the module, so when we access a class via the path, it should always be possible to find the defined `defn` field.

LPTK replied:
@ChingLongTin are you going to follow @NeilKleistGao's suggestion in this PR? Or in a follow-up one?

ChingLongTin replied:
I will do it in a follow-up one.

ChingLongTin committed:
Update hkmc2/shared/src/main/scala/hkmc2/codegen/Instrumentation.scala

Co-authored-by: Fa1sePRoMiSe <neilkleist@gmail.com>

ChingLongTin committed:
replace ??? with ErrorReport

ChingLongTin committed:
Merge branch 'instrument' of https://github.com/ChingLongTin/mlscript into instrument

ChingLongTin committed:
cleanup

ChingLongTin committed:
remove some crashes from .get

ChingLongTin committed:
rename transformFunDefn

ChingLongTin committed:
remove inf parameter form Shape.Arr and Block.Tup

ChingLongTin committed:
remove dead code

ChingLongTin committed:
Merge remote-tracking branch 'mlscript/hkmc2' into instrument

ChingLongTin committed:
use :todo for spread test case

ChingLongTin commented:
There's a diff test change on `backlog/NamedArgs.mls`, `codegen/BasicTerms.mls`, `codegen/Do.mls`, `ucs/syntax/NestedOpSplits.mls`, from the printing of the `Predef` symbol. For example in the first file we have the following diff:
```
diff --git a/hkmc2/shared/src/test/mlscript/backlog/NamedArgs.mls b/hkmc2/shared/src/test/mlscript/backlog/NamedArgs.mls
index e5e0724..4c99118 100644
--- a/hkmc2/shared/src/test/mlscript/backlog/NamedArgs.mls
+++ b/hkmc2/shared/src/test/mlscript/backlog/NamedArgs.mls
@@ -37,7 +37,7 @@ id(a: 0)
 //│   stats = Nil
 //│   res = App:
 //│     lhs = SynthSel{sym=member:id⁰}:
-//│       prefix = Ref{sym=member:Predef⁰} of member:Predef⁰
+//│       prefix = Ref{sym=member:Predef¹} of member:Predef¹
 //│       nme = Ident of "id"
 //│     rhs = Tup of Ls of 
 //│       Fld:
```

LPTK committed:
Merge branch 'hkmc2' into instrument

LPTK commented:
@ChingLongTin, indeed, there was some flakiness that should be fixed now.

ChingLongTin committed:
Merge branch 'hkmc2' into instrument

LPTK committed:
Merge branch 'hkmc2' into instrument

LPTK committed:
Merge branch 'hkmc2' into instrument

LPTK committed:
Update difftests

LPTK commented on hkmc2/shared/src/test/mlscript/block-staging/Functions.mls:
```diff
@@ -0,0 +1,135 @@
+//│ > FunDefn(Symbol("tup2"), ([]), Return(Tuple([Lit(1), Ref(Symbol("x1"))]), false), true)
+//│ > FunDefn(Symbol("dynsel"), ([]), Scoped([Symbol("tmp")], Assign(Symbol("tmp"), Tuple([Lit(1)]), Return(DynSelect(Ref(Symbol("tmp")), Lit(0), false), false))), true)
+//│ > FunDefn(Symbol("match1"), ([]), Scoped([Symbol("scrut")], Assign(Symbol("scrut"), Lit(9), Match(Ref(Symbol("scrut")), [Cls(ClassSymbol("Bool"), Select(Ref(Symbol("runtime")), Symbol("unreachable"))) -> Return(Lit(1), false), Lit(8) -> Return(Lit(2), false), Cls(ClassSymbol("Int"), Select(Ref(Symbol("runtime")), Symbol("unreachable"))) -> Return(Lit(3), false)], Return(Lit(0), false), End))) ...
+//│ > FunDefn(Symbol("match2"), ([]), Scoped([Symbol("tmp1"), Symbol("element1$"), Symbol("element0$"), Symbol("a")], Label(Symbol("split_root$"), false, Label(Symbol("split_1$"), false, Match(Ref(Symbol("x1")), [Tup(0) -> Assign(Symbol("tmp1"), Lit(1), Break(Symbol("split_root$"))), Tup(2) -> Assign(Symbol("element0$"), Call(Select(Select(Ref(Symbol("runtime")), Symbol("Tuple")), Symbol("get")), ...
```
I just noticed these suspicious `NaN` occurrences. Is a wrong concatenation being done?

LPTK replied:
(The problem is still there after the merge.)

ChingLongTin replied:
oops, i fixed it now.

ChingLongTin committed:
fix printing bug

LPTK reviewed (approved):
Ok, LGTM, but please remember to address the two remaining comments in follow-up PRs.
